indexing-service/src/main/java/org/apache/druid/indexing/input/DruidSegmentReader.java

```diff
@@ -0,0 +1,225 @@
+            storageAdapter.getAdapter().makeCursors(
+                Filters.toFilter(dimFilter),
+                storageAdapter.getInterval(),
+                VirtualColumns.EMPTY,
+                Granularities.ALL,
+                false,
+                null
+            ),
+            cursor -> {
+              final BaseLongColumnValueSelector timestampColumnSelector =
+                  cursor.getColumnSelectorFactory().makeColumnValueSelector(ColumnHolder.TIME_COLUMN_NAME);
+
+              final Map<String, DimensionSelector> dimSelectors = new HashMap<>();
+              for (String dim : dimensions) {
+                final DimensionSelector dimSelector = cursor
+                    .getColumnSelectorFactory()
+                    .makeDimensionSelector(new DefaultDimensionSpec(dim, dim));
+                // dimSelector is null if the dimension is not present
+                if (dimSelector != null) {
+                  dimSelectors.put(dim, dimSelector);
+                }
+              }
+
+              final Map<String, BaseObjectColumnValueSelector> metSelectors = new HashMap<>();
+              for (String metric : metrics) {
+                final BaseObjectColumnValueSelector metricSelector =
+                    cursor.getColumnSelectorFactory().makeColumnValueSelector(metric);
+                metSelectors.put(metric, metricSelector);
+              }
+
+              return Sequences.simple(
+                  () -> new Iterator<Map<String, Object>>()
+                  {
+                    @Override
+                    public boolean hasNext()
+                    {
+                      return !cursor.isDone();
+                    }
+
+                    @Override
+                    public Map<String, Object> next()
+                    {
+                      final Map<String, Object> theEvent = Maps.newLinkedHashMap();
+                      final long timestamp = timestampColumnSelector.getLong();
+                      theEvent.put(TimestampSpec.DEFAULT_COLUMN, DateTimes.utc(timestamp));
+
+                      for (Entry<String, DimensionSelector> dimSelector : dimSelectors.entrySet()) {
+                        final String dim = dimSelector.getKey();
+                        final DimensionSelector selector = dimSelector.getValue();
+                        final IndexedInts vals = selector.getRow();
+
+                        int valsSize = vals.size();
+                        if (valsSize == 1) {
+                          final String dimVal = selector.lookupName(vals.get(0));
+                          theEvent.put(dim, dimVal);
+                        } else if (valsSize > 1) {
+                          List<String> dimVals = new ArrayList<>(valsSize);
+                          for (int i = 0; i < valsSize; ++i) {
+                            dimVals.add(selector.lookupName(vals.get(i)));
+                          }
+                          theEvent.put(dim, dimVals);
```

Comment:
This part does not seem to be covered by any of the tests in `org.apache.druid.indexing`

Reply:
I adjusted `CompactionRunTaskTest` so that there's a multivalued dimension (this case handles that), and added a row correctness check to `testRun` there